### PR TITLE
Preserve UTF-8 bytes when generating JSON files

### PR DESCRIPTION
**Pull request by muojp**
Hi. Here's a fix for #356.
In the original code, non-ASCII characters got broken when writing to files. I put a proxy function to keep bytes as they were.

Then, I tested the code against PowerShell 5.1 / 6.2 (currently under preview: https://github.com/PowerShell/PowerShell/releases) and found `-Encoding Byte` option was gone. So, I also put a version branch to keep forward-compatibility.

Tested on PowerShell 5.1 (Windows 10) and 6.x (Windows/Linux).

### CODE REVIEW

**Comment by ashmind** [comment]
Thanks for submitting this! I'm not actively working on SharpLab, so my replies are somewhat delayed.
I'll review now.

File: #build/Publish-All.ps1 [hunk 1]
```diff
@@ -79,6 +79,15 @@ function Get-PredefinedBranches() {
     })
 }
 
+function Set-Content-UTF8($file, $content) {
+    if ($HOST.Version.Major -ge 6) {
+        # PowerShell Core 6.0+ natively support UTF-8 output but no more -Encoding Byte support
+        return Set-Content $file $content
+    } else {
+        return Set-Content $file -Encoding Byte ([Text.Encoding]::UTF8.GetBytes($content))
```
Comment (ashmind) [inline #build/Publish-All.ps1]: Can this use Set-Content -Encoding UTF8 (and is that still supported in PowerShell 6)?

File: #build/Publish-All.ps1 [hunk 2]
```diff
@@ -126,7 +135,7 @@ try {
         Write-Output "*** $_"
 
         $siteRoslynArtifactsRoot = Resolve-Path "$roslynArtifactsRoot\$($_.Name)"
-        $branchInfo = ConvertFrom-Json ([IO.File]::ReadAllText("$siteRoslynArtifactsRoot\BranchInfo.json"))
+        $branchInfo = ConvertFrom-Json ([IO.File]::ReadAllText((Resolve-Path "$siteRoslynArtifactsRoot\BranchInfo.json")))
```
Comment (ashmind) [inline #build/Publish-All.ps1]: `$siteRoslynArtifactsRoot` should already be resolved by this point though?

File: #build/Publish-All.ps1 [hunk 3]
```diff
@@ -79,6 +79,15 @@ function Get-PredefinedBranches() {
     })
 }
 
+function Set-Content-UTF8($file, $content) {
```
Comment (ashmind) [inline #build/Publish-All.ps1]: I would prefer `Set-Utf8Content` or `Set-ContentUtf8` - I think PowerShell naming guidelines just use one dash.

**Comment by muojp** [comment]
Hi. Thank you for taking time to review the changes!

File: #build/Publish-All.ps1 (hunk 1 above)
Comment (muojp) [reply #build/Publish-All.ps1]: `Set-Content -Encoding UTF8` is still supported in PowerShell 6, and it's equivalent to `Set-Content -Encoding` which write out UTF8N string.
But `Set-Content -Encoding UTF8` in PowerShell -5 generates UTF8 (w/ BOM), which I assume is not suitable for modern Web. My assumption might be incorrect.

File: #build/Publish-All.ps1 (hunk 2 above)
Comment (muojp) [reply #build/Publish-All.ps1]: Good point! It's related to how PowerShell 6 handles `\`.
Linux/macOS versions of PowerShell 6 doesn't interpret `\` as a path-separator unless used in `Resolve-Path`, so the original code raise error (FileNotFound) in Linux/macOS. Unfortunately, we need to `Resolve-Path`-ize things like we call Path.Combine() in C# for cross-platform.

File: #build/Publish-All.ps1 (hunk 3 above)
Comment (muojp) [reply #build/Publish-All.ps1]: Thank you! Pushed a fix.

**Comment by ashmind** [comment]
Thanks for your patience!
I applied some more changes -- found that build phase needed a similar fix, and simplified the function (inlined as `-Encoding UTF8`).

**Pull request merged**